### PR TITLE
Initial Bazel Groundwork

### PR DESCRIPTION
**Pull request by ejfitzgerald**
This PR is largely based on @troelsfr work and lays the ground work for switching to a bazel based build in the future.

There are a number of issues with the bazel build as it current stands, however, the principal reason for merging at this point is in order to reintegrate the current structural changes that have been necessary in order to remove circular references

*Update* Bazel configuration removed from the project at this time and will be added back at a later point. Some addition work is still required here

### CODE REVIEW

File: apps/constellation/constellation.cpp [hunk 1]
```diff
@@ -16,8 +16,8 @@
 //
 //------------------------------------------------------------------------------
 
+#include "bloom_filter/bloom_filter.hpp"
```
Comment (n-hutton) [inline apps/constellation/constellation.cpp]: I am a bit confused - how did this relate to bazel?

File: bazel/fetchai_build_system.bzl [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    # since we are emulating the current structure we need to provide additional include
+    # paths based on the dependencies
+    for dep in deps:
+        if dep.startswith('//libs/'):
+            _, _, dep_name = dep.rpartition(':')
+            include_ops.append("-Ilibs/" + dep_name + "/include")
+        elif dep == '//vendor:spdlog':
+            include_ops.append("-Ivendor/spdlog/include")
+        elif dep == '//vendor:backward-cpp':
+            include_ops.append("-Ivendor/backward-cpp")
```
Comment (n-hutton) [inline bazel/fetchai_build_system.bzl]: Troels mentions that backward cpp might be made obsolete by bazil due to a limitation that means you can't get debug symbols in your binary - I suggest that you delete it if it is causing some pain and I can put it back in at a later date if we want it

File: bazel/fetchai_build_system.bzl [hunk 3]
```diff
@@ -0,0 +1,121 @@
+load(
+    ":fetchai_copts.bzl",
+    "fetchai_copts",
+)
+
+
+def fetchai_cc_deps(deps):
+    include_ops = []
+
+    # since we are emulating the current structure we need to provide additional include
```
Comment (n-hutton) [inline bazel/fetchai_build_system.bzl]: 👍

File: libs/address/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+# Main Library Target
+# ------------------------------------------------------------------------------
+
+setup_library(fetch-address)
+target_link_libraries(fetch-address PUBLIC fetch-core fetch-crypto fetch-logging)
+# ------------------------------------------------------------------------------
+# Example Targets
+# ------------------------------------------------------------------------------
+
+# add_subdirectory(examples)
```
Comment (n-hutton) [inline libs/address/CMakeLists.txt]: dead?

File: libs/core/include/core/byte_array/const_byte_array.hpp [hunk 5]
```diff
@@ -595,3 +584,29 @@ ConstByteArray operator+(char const *a, ConstByteArray const &b);
 
 }  // namespace byte_array
 }  // namespace fetch
+
+namespace std {
```
Comment (n-hutton) [inline libs/core/include/core/byte_array/const_byte_array.hpp:599]: 👍  - as an aside, it might be cleaner to do the method where you pass in a class to maps as a template argument

File: libs/core/src/reactor.cpp [hunk 6]
```diff
@@ -180,12 +179,14 @@ void Reactor::Monitor()
       }
       catch (std::exception const &ex)
       {
-        FETCH_LOG_WARN(LOGGING_NAME, "The reactor caught an exception! ", name_,
-                       " error: ", ex.what());
+        // TODO(tfr): make logging indendent of core        FETCH_LOG_WARN(LOGGING_NAME, "The
```
Comment (n-hutton) [inline libs/core/src/reactor.cpp]: ah, is this to break dependencies?

File: libs/json/tests/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+project(fetch-json)
+
+# CMake configuration
+include(${FETCH_ROOT_CMAKE_DIR}/BuildTools.cmake)
+
+# Compiler Configuration
+setup_compiler()
+
+fetch_add_test(json_gtest fetch-core json/gtest/)
+# fetch_add_test(yaml_gtest fetch-core yaml/gtest/)
```
Comment (n-hutton) [inline libs/json/tests/CMakeLists.txt]: delete line?

**Review by n-hutton — DISMISSED** [review]
Looking good - I don't know how the bazel system works so I can't comment much on that

File: apps/constellation/constellation.cpp (hunk 1 above)
Comment (troels-im) [reply apps/constellation/constellation.cpp]: It does not directly, but it does in that bloom_filter introduces a circular dependency and part of this PR is to get rid of those.

File: bazel/fetchai_build_system.bzl (hunk 2 above)
Comment (troels-im) [reply bazel/fetchai_build_system.bzl]: I did not say that `backward-cpp` was obsolete, just that it was uncertain whether it would have any impact on `backward-cpp` and that we might need to have a special production compile setting.

File: libs/core/include/core/byte_array/const_byte_array.hpp (hunk 5 above)
Comment (troels-im) [reply libs/core/include/core/byte_array/const_byte_array.hpp:599]: This would have been my preferred option, but at some point, this was introduced to the code and since then we have a had it used in a number of places. It is another PR to solve this and high priority atm.

File: libs/core/src/reactor.cpp (hunk 6 above)
Comment (troels-im) [reply libs/core/src/reactor.cpp]: Yes, logging is now it's own module that depends on `core`. This means that anything in core cannot use it atm. One solution would be get remove `ConstByteArray` from logging - I believe this was the only reason why it depends on `core`.

File: libs/address/CMakeLists.txt (hunk 4 above)
Comment (ejfitzgerald) [reply libs/address/CMakeLists.txt]: done

File: libs/json/tests/CMakeLists.txt (hunk 7 above)
Comment (ejfitzgerald) [reply libs/json/tests/CMakeLists.txt]: Done

File: libs/crypto/src/sha256.cpp [hunk 8]
```diff
@@ -22,6 +22,7 @@
 #include <cassert>
 #include <cstddef>
 #include <cstdint>
+#include <openssl/evp.h>
```
Comment (troels-im) [inline libs/crypto/src/sha256.cpp]: Is this one still needed?

Comment (ejfitzgerald) [reply libs/crypto/src/sha256.cpp]: Nope good spot

File: libs/ledger/include/ledger/chaincode/wallet_record.hpp [hunk 9]
```diff
@@ -19,15 +19,16 @@
 
 #include "core/byte_array/decoders.hpp"
 #include "ledger/chaincode/deed.hpp"
+#include "logging/logging.hpp"
 
 #include <cstdint>
 #include <map>
 #include <memory>
 
 // TODO(HUT): doesn't putting this here pollute the namespace?
```
Comment (troels-im) [inline libs/ledger/include/ledger/chaincode/wallet_record.hpp]: Could we either solve this or add an issue number?

File: libs/ledger/include/ledger/chaincode/wallet_record.hpp [hunk 10]
```diff
@@ -19,15 +19,16 @@
...
-using fetch::variant::Variant;
 using fetch::byte_array::ConstByteArray;
 using fetch::byte_array::FromBase64;
+using fetch::variant::Variant;
```
Comment (troels-im) [inline libs/ledger/include/ledger/chaincode/wallet_record.hpp]: And yes, this seems to pollute the global namespace.

File: libs/network/CMakeLists.txt [hunk 11]
```diff
@@ -19,8 +19,10 @@ target_link_libraries(fetch-network
                       PUBLIC fetch-core
                              fetch-crypto
                              fetch-math
-                             fetch-muddle # remove
+                             fetch-json
+                             fetch-muddle # TODO: remove
```
Comment (troels-im) [inline libs/network/CMakeLists.txt]: TODO: Remove

File: scripts/create_pb_wrappers.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
+import os
+import glob
```
Comment (troels-im) [inline scripts/create_pb_wrappers.py]: This file seems irrelevant to this PR.

**Review by troels-im — DISMISSED** [review]
LGTM. Only a few minor comments - either make issues or fix as you see fit.

File: scripts/create_pb_wrappers.py (hunk 12 above)
Comment (ejfitzgerald) [reply scripts/create_pb_wrappers.py]: Fixed

File: libs/network/CMakeLists.txt (hunk 11 above)
Comment (ejfitzgerald) [reply libs/network/CMakeLists.txt]: Fixed

File: libs/ledger/include/ledger/chaincode/wallet_record.hpp (hunk 10 above)
Comment (ejfitzgerald) [reply libs/ledger/include/ledger/chaincode/wallet_record.hpp]: Fixed

File: libs/ledger/include/ledger/chaincode/wallet_record.hpp (hunk 9 above)
Comment (ejfitzgerald) [reply libs/ledger/include/ledger/chaincode/wallet_record.hpp]: Fixed

**Review by troels-im — DISMISSED** [review]

**Review by troels-im — DISMISSED** [review]
LGTM

**Review by n-hutton — APPROVED** [review]
